mcodingbot/plugins/highlights.py

```diff
@@ -0,0 +1,179 @@
+
+    async def callback(self, ctx: Context) -> None:
+        highlight = await Highlight.exists(highlight=self.word)
+
+        was_deleted = False
+        if highlight:
+            deleted_highlights = (
+                await UserHighlight.delete_query()
+                .where(highlight_id=highlight.id, user_id=ctx.user.id)
+                .execute()
+            )
+            was_deleted = bool(len(deleted_highlights))
+
+        if was_deleted:
+            _uncache_highlight(self.word, ctx.user.id)
+            await ctx.respond(f'Removed "{self.word}" from your highlights.')
+            return
+
+        await ctx.respond(
+            f'"{self.word}" was not one of your highlights.', ephemeral=True
+        )
+
+
+@plugin.include
+@highlights_group.child
+@crescent.command(name="list", description="List all of your highlights.")
+async def list(ctx: Context) -> None:
+    user = await User.exists(user_id=ctx.user.id)
+
+    if user is None:
+        await ctx.respond("You do not have any highlights.")
+        return
+
+    highlights = await user.highlights.fetchmany()
+
+    if not highlights:
+        await ctx.respond("You do not have any highlights.")
+        return
+
+    embed = hikari.Embed(
+        title="Your Highlights",
+        description="\n".join(hl.highlight for hl in highlights),
+        color=CONFIG.theme,
+    )
+    await ctx.respond(embed=embed)
+
+
+@plugin.include
+@crescent.event
+async def on_start(_: hikari.StartingEvent) -> None:
+    highlights = await Highlight.fetchmany()
+
+    for highlight, user_highlights in zip(
+        highlights,
+        await asyncio.gather(
+            *(
+                UserHighlight.fetchmany(highlight_id=highlight.id)
+                for highlight in highlights
+            )
+        ),
+    ):
+        _cache_highlight(
+            highlight.highlight,
+            *(user_highlight.user_id for user_highlight in user_highlights),
+        )
+
+
+async def _dm_user_highlight(
+    triggering_message: hikari.Message, trigger: str, user_id: int
+) -> None:
+    _avatar_url = triggering_message.author.avatar_url
+    avatar_url = _avatar_url.url if _avatar_url else None
+    embed = (
+        hikari.Embed(
+            title="Highlight Triggered",
+            url=triggering_message.make_link(CONFIG.mcoding_server),
+            description=triggering_message.content,
+            color=CONFIG.theme,
+        )
+        .set_footer(f"Highlight: {trigger}")
+        .set_author(name=triggering_message.author.username, icon=avatar_url)
+    )
+    channel = await plugin.app.rest.create_dm_channel(user_id)
+    await channel.send(embed=embed)
+
+
+@plugin.include
+@crescent.event
+async def on_message(event: hikari.GuildMessageCreateEvent) -> None:
+    if not event.content or event.is_bot:
+        return
+
+    for highlight, users in highlights_cache.items():
+        if highlight in event.content.split():
```

Comment:
Doing this means that if the word is followed by a comma, it will not trigger. I would just go with `highlight in event.content` for now. Regex support in the future anyways, I think?

Reply:
I would prefer stripping commas before comparing

Reply:
If regex won't be available now then some strings need to be updated to reflect that.

Reply:
> I would prefer stripping commas before comparing

Any non-alphanumerica characters should be stripped. Here's a list I can think of: ```
`'";:,.<>?/\|[]{}()*&^%$@!
```

The problem is, this means that highlights that rely on non-alphanumeric characters won't work right. I really think we should just remove the split.

Reply:
> If regex won't be available now then some strings need to be updated to reflect that.

I didn't see any, maybe go through and add comments to those?

Reply:
> I didn't see any, maybe go through and add comments to those?


I see them. I can go fix it.

Reply:
> The problem is, this means that highlights that rely on non-alphanumeric characters won't work right. I really think we should just remove the split.

It might just be less work to add regex support now

Reply:
On second thought we should definitely remove the split, otherwise multi-word highlights don't work.

Reply:
> > I would prefer stripping commas before comparing
> 
> Any non-alphanumerica characters should be stripped. Here's a list I can think of:
> 
> ```
> `'";:,.<>?/\|[]{}()*&^%$@!
> ```

So just `string.punctuation` maybe?

Reply:
```pycon
>>> import string
>>> print(string.punctuation)
!"#$%&'()*+,-./:;<=>?@[\]^_`{|}~
```

Reply:
@trag1c very helpful thanks.

Still, that leaves us with the problem of having highlights that include punctuation.

Reply:
If we don't want to deal with implementing the regex timeout code now we could use escape the highlight and then wrap it in `\b`s, then match on that with regex. It's a simple enough pattern that it shouldn't have performance issues.

Reply:
> Still, that leaves us with the problem of having highlights that include punctuation.

or (and more importantly) spaces.

Reply:
> If we don't want to deal with implementing the regex timeout code now we could use escape the highlight and then wrap it in \bs, then match on that with regex. It's a simple enough pattern that it shouldn't have performance issues.

won't that just leave us back with the problem of `test` not matchint `test.`?

Reply:
> won't that just leave us back with the problem of `test` not matchint `test.`?

I'm pretty sure that between a letter and a period counts as a word boundary, so the regex `\btest\b` (which corresponds to the highlight `test`) should match `test.`.

The only issue I can see with this is properly escaping the input string.

Reply:
All of this will be solved once we have regex highlights I think. But for now, I think using `highlight in message.content` allows more control over what's matched and what isn't than doing `highlight in message.content.split()`. Neither are completely ideal, but again, we can fix that with regex highlights.

Reply:
Doing just `in` for now is the easiest I guess

Reply:
Sounds good (test comment)